scipy/stats/_multivariate.py

```diff
@@ -636,6 +638,30 @@ def cdf(self, x, mean=None, cov=1, allow_singular=False, maxpts=None,
         out = self._cdf(x, mean, cov, maxpts, abseps, releps)
         return out
 
+    def _rvs(self, mean, cov_L, size, random_state=None):
+        """
+        Parameters
+        ----------
+        mean : ndarray
+            Mean of the distribution
+        cov_L : ndarray
+            A decomposition such that np.dot(cov_L, cov_L.T)
+            is the covariance matrix.
+        size : integer
+            Number of samples to draw.
+        %(_doc_random_state)s
+
+        Notes
+        -----
+        As this function does no argument checking, it should not be
+        called directly; use 'rvs' instead.
+
+        """
+        random_state = self._get_random_state(random_state)
+        std_norm = random_state.standard_normal((size, mean.size))
+        out = mean + np.dot(std_norm, cov_L.T)
```

Comment:
Can you point me to a reference that you can get 
Is this what goes on under the hood in NumPy's `multivariate_normal.rvs`?

Reply:
Not quite: `U` is such that `np.dot(U, U.T)` equals the inverse covariance matrix, while `L` is such that `np.dot(L, L.T)` equals the covariance matrix. As far as I can tell, `np.sqrt(s)[:, None] * v` in that Numpy snippet is equivalent to `L.T` here.

Reply:
OK, I see how it compares to the note in Wikipedia. Any reason to prefer `np.dot(std_norm, cov_L.T)` over `cov_L @ std_norm`?

Reply:
That `np.sqrt(s)[:, None] * v` is equivalent in the sense that if `B.T = np.sqrt(s)[:, None] * v`, then `B @ B.T == cov`. But they're not the same matrix. The elements have the same magnitudes, but I'm not familiar with the relationship. I wrote this to help me understand:
```
import numpy as np
from numpy.linalg import svd

cov = np.array([[4, 1, 0.5], [1, 3, 0.9], [0.5, 0.9, 2]])
mean = np.array([1, 2, 3])
size = (3, 4)
rng = np.random.default_rng(0)

L, S = np.linalg.eigh(cov)
A = S * np.sqrt(L)
x = rng.normal(size=(*size, len(A)))
rvs_sp = mean + x @ A.T

(u, s, v) = svd(cov)
B = np.sqrt(s)[:, None] * v
rvs_np = mean + x @ B

print(A)
print(B)
```
prints 
```
[[ 0.04779366 -1.01874786 -1.72042686]
 [-0.63598445  1.05083947 -1.22117164]
 [ 1.02988709  0.69619987 -0.67426874]]
[[-1.72042686 -1.22117164 -0.67426874]
 [ 1.01874786 -1.05083947 -0.69619987]
 [ 0.04779366 -0.63598445  1.02988709]]
```
I would suggest that you just use `svd`, but it looks like a `_PSD` is intantiated and `eigh` is calculated when the distribution is frozen? So I see why you wouldn't want to _also_ do `svd` if `eigh` is already done. I imagine there is some way to transform one into the other so you can keep compatibility with previous `multivariate_normal` RVS?

Reply:
Mathematically, they should be equivalent. The difference is due to two reasons:

i) `eigh` orders eigenvalues in ascending order, while `svd` uses descending order.
ii) Because `eigh` and `svd` use different underlying LAPACK functions (`_syevd`, `_dgesdd`), they can arbitrarily flip the sign of an eigen/singular vector.  The factorizations remain consistent, but become non-unique up to the signs of the vectors. I need to read up more to figure out whether these sign flips can be reliably predicted or not.  

Off the top of my head, one way to maintain backwards compatibility  is to substitute`eigh` with `svd` entirely in `_PSD`. Although this takes care of the sample problem, `_PSD.U` is now different. This doesn't break anything mathematically, functions like `logpdf`, `pdf`, and `pinv` will remain identical (up to numerical round-off) to what they were returning before. But it'll spell trouble for anyone directly relying on the `U` attribute.

Here's some more information on the sign flip problem: https://prod-ng.sandia.gov/techlib-noauth/access-control.cgi/2007/076422.pdf

Reply:
Ok. Well, it looks like `_PSD` is private, so `U` might not be exposed to users anywhere.
Maybe you could add an argument to the `_PSD` initializer that allows you to create a `_PSD` object that uses SVD internally instead eigh? The default could be to use eigh like it does now, but the multivariate_normal code could pass the argument that makes it use SVD. Does that make sense?

Reply:
I'm sorry to continue dragging this, but as I was playing around with this today, I realized that `eigh` is much faster than `svd`. For example:

```python
[nav] In [53]: dim = 2000; cov = wishart.rvs(dim + 1, scale=np.diag(np.ran
          ...: dom.rand(dim)))

[nav] In [54]: %timeit u, s, v = scipy.linalg.svd(cov)
1.6 s ± 30.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

[ins] In [55]: %timeit ss, vv = scipy.linalg.eigh(cov, driver='evd')
485 ms ± 4.57 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

I think there's a way to leverage the faster execution of `eigh`, while still maintaining backwards compatibility. It'll involve adding a `method = {'svd', 'eigh'}` argument to:

* all methods of `multivariate_normal_gen`, including `__call__`. For instance:
    * `logpdf(x, mean=None, cov=1, method='svd', allow_singular=False)`
    *  `rvs(mean=None, cov=1, size=1, method='svd', random_state=None)`,
    * `__call__(self, mean=None, cov=1, method='svd', allow_singular=False, seed=None)`
*  the `multivariate_normal_frozen` constructor
* the `_PSD` constructor

The `method` kwarg in `multivariate_normal_gen` and `multivariate_normal_frozen` will get passed to the `_PSD` constructor, which'll use the appropriate algorithm to compute the factorization. 

Because we default to `method='svd'`, samples remain backwards compatible with Numpy, but new code can use `method='eigh'` for faster execution. In fact, this also leaves open the possibility that `method='cholesky'` can be added later (this is faster than `eigh`, though less robust, as noted [here](https://github.com/numpy/numpy/blob/7a505741392ee24cb85ee096eec3b3e4a624ac15/numpy/random/_generator.pyx#L3458)). 

Not only does it retain backwards compatibility, `method='eigh'` will produce identical samples as `numpy.random.Generator.multivariate_normal(..., method='eigh')` (that's why `driver='evd'` was necessary in `scipy.linalg.eigh`).

There is a disadvantage though. Setting `method='svd'` as the default *everywhere* will slow down extant code. That's easily fixed by setting `method='eigh'` as the default in `pdf`, `logpdf` etc., but `method='svd'` in `rvs`, which potentially looks ugly. 

This slowdown will also happen with all methods of  `multivariate_normal_frozen`, except `rvs`, if follow the approach we came up with (in the couple of comments right above this one).

What do you think @mdhaber?

Reply:
If you have a quasi-guarantee that the cov matrix is sufficiently and numerically positive definite you don't need svd or eigh; cholesky is the definite method. The issue gets slightly complicated when covariance matrix gets too close to positive semidefiniteness then `cholesky` might fail even though the matrix is theoretically pos.def.

Note that `svd` assumes no structure on the matrix not even squareness, `eigh` assumes just hermitianness, cholesky demands positive definiteness, hence the speedups. So if positive definiteness is given then anything other than `cholesky` is a waste.

Typically you attempt a cholesky first and if it fails you fall back to eigh as a backup, and since cholesky is much cheaper this strategy is beneficial when semidefiniteness is rare.

Reply:
@siddhantwahal I understand the motivation, and I think it would be nice to be able to select the method, but is there precedent for any stats distribution methods having distribution-specific parameters other than shape parameters?

If so, go for it. If not, consider making the change internally - `rvs` will use `svd`, the other methods will use `eigh` as they do now. In either case, I think `eigh` should stay the default, and `svd` would be the default only in `rvs`. A compromise for frozen distributions might be to stick with `eigh` until `rvs` is called: the first time `rvs` is called the matrix would be re-factorized with `svd`, and that would be stored.

Reply:
Apologies for the delayed response, this was a busy week.

I couldn't find find a precedent for stats methods accepting arguments pertaining to the algorithm used (with the exception of numpy, of course). 

> In any case, I think eigh should stay the default for most methods, and svd would be the default only in rvs. A compromise for frozen distributions might be to use the factorization provided by eigh until rvs is called for the first time, at which point the the matrix would be re-factorized with svd, and that factorization would be stored for future use.

That makes sense. Thanks for the feedback!